Implement BIP78 payjoin in JoinMarketQt GUI

kristapsk opened the pull request:
<del>Not finished, but I was able to successfully do a testnet payjoin from bech32 JM wallet to Kukks payjoin test site.</del>

I tried to follow the UX Wasabi have, just instead of adding dynamically adding another input box when the BIP78 URI is pasted in address field, Payjoin endpoint field replaces Number of counterparties field. You can get back to JoinMarket coinjoins mode by deleting that URI and pressing Tab or clicking on another field.

Can be tested and any feedback is welcome.

AdamISZ commented on scripts/joinmarket-qt.py:
```diff
@@ -698,6 +739,15 @@ def qt_directsend_callback(rtxd, rtxid, confs):
+            reactor.callWhenRunning(send_payjoin, manager,
+                accept_callback=self.checkDirectSend,
+                info_callback=self.infoDirectSend)
+            reactor.run()
```
This won't work as it leads to `reactorAlreadyRunning` errors (for JMQT the global twisted reactor is started in `reactor.runReturn()` at the end of the file, at the start of the process). To fire a call asynchronously and then continue, use `task.deferLater(0.0, function_name, args..)` (we use it in quite a few places) instead of `callWhenRunning`.

AdamISZ commented:
This is good, I'm hoping we can build the 0.7.0 release once this is merged, as it'll be the completion of the sender-side BIP78 work.

Something I observed in testing, which is test specific:

You may or may not have noticed that in `jmclient.payjoin.py` there is a `WhitelistContextFactory` object that is [used](https://github.com/JoinMarket-Org/joinmarket-clientserver/blob/master/jmclient/jmclient/payjoin.py#L429) as an optional parameter to `send_payjoin`, and this is only used for tests, as it allows SSL with a localhost server that doesn't have a certificate. See `test/payjoinclient.py` and corresponding `/test/payjoinserver.py`.

So, to test this in a regtest environment, I had to make one additional tweak (apart from the `task.deferLater` thing mentioned above): I had to add another argument `tls_whitelist=[b"127.0.0.1"]`. So overall I had:

```python3
        if bip78url:
            tls_whitelist_for_testing = [b"127.0.0.1"]
            manager = parse_payjoin_setup(self.bip21_uri,
                mainWindow.wallet_service, mixdepth)
            task.deferLater(reactor, 0.0, send_payjoin, manager,
                            accept_callback=self.checkDirectSend,
                            info_callback=self.infoDirectSend,
                            tls_whitelist=tls_whitelist_for_testing)
            return
```
With this change, I was able to successfully test: loading a uri into the address/uri field, seeing the fields populate correctly, pressing start, and having the payjoin go through successfully. By the way the server side I have running with the command:

```bash
pytest --btcconf=/path/to/bitcoin.conf --btcroot=/path/to/bitcoin/bin/ --btcpwd=123456abcdef --nirc=2 test/payjoinserver.py -s -p no:warning
```

I confirm also that deleting the URI and pressing tab or leaving, reverts to former address field. But I guess you agree we need a slightly better version of that, as the user could be quite confused.

AdamISZ commented on scripts/joinmarket-qt.py:
```diff
@@ -103,6 +104,22 @@ def update_config_for_gui():
             jm_single().config.set("GUI", gcn, gcv)
 
 
+def switchToBIP78Payjoin(parent, endpoint_url):
```
Are these two functions not more naturally methods of the Spend tab?

kristapsk replied:
Yes, together with `checkAddress()`.

kristapsk replied:
And 'checkAmount()` too.

kristapsk commented:
> But I guess you agree we need a slightly better version of that, as the user could be quite confused.

Not sure. I was thinking about better tooltips and docs. Probably some delete/close type icon could be added at the right side of that input box?

AdamISZ commented:
Tbh we should probably take the deferred that is returned by `task.deferLater` and add a callback to it, that wipes the fields; and also disable the fields while the payment is running. I'll try to mock something up later today and see if it more or less works, if so i'll write the code here.

AdamISZ commented:
Another problem encountered is that frequently I get this error when pressing `Start`: 

```
Traceback (most recent call last):
  File "joinmarket-qt.py", line 747, in startSingle
    mainWindow.wallet_service, mixdepth)
  File "/home/waxwing/testjminstall/joinmarket-clientserver/jmclient/jmclient/payjoin.py", line 380, in parse_payjoin_setup
    assert btc.is_bip21_uri(bip21_uri), "invalid bip21 uri: " + bip21_uri
TypeError: must be str, not NoneType
```

.. I notice that this code block is being reached on some kind of second-pass through call of `checkAddress()`: https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/659/files#diff-3a1ec99e5c3829c412ea7b81a7b15be7R143

(because I added print statements). I'm not sure why it calls this function twice, once when I first leave the recipient input box, and second time when i click start. that second time, it is not of course a bip21 uri (it's an address) and bi21_uri is set to None, causing the error. I'll leave this one with you :)

AdamISZ commented:
Another problem I have found is that `jmclient.payjoin.py` will need to be refactored a bit, for two reasons.

First, for context, note there are a chain of callbacks: `send_payjoin -> receive_payjoin_proposal_from_server -> process_payjoin_proposal_from_server`. Note that at the end of that last function, `reactor.stop()` is called.

So that's the first thing that must change: stopping the reactor is the right choice to end the program for the command line case, but should be omitted for Qt via some kind of argument/flag.

The second problem is the fallback. In case of errors we want to run the fallback transaction code in `fallback_nonpayjoin_broadcast` but again, for Qt, we don't want to stop the reactor, so the same comment about arg/flag.

But even more: in case of *success*, the fallback code still executes: because of this line: https://github.com/JoinMarket-Org/joinmarket-clientserver/blob/692ff4decadfe1a36ca30f8a3d983d6337acd241/jmclient/jmclient/payjoin.py#L453 , which is part of the BIP78 spec, we will always *try* to broadcast the transaction after 60 seconds, in case some unrecognised errors occurred. So right now, after I successfully do a payjoin in Qt, after 60 seconds I see loud and incorrect warnings about "payment was NOT made", because the fallback fails due to missing inputs.
Our best choice is probably to explicitly cancel this delayed call in case of success (other options will be confusing).

AdamISZ commented:
Wrt that last comment, I'll make a short PR that makes those actions optional, then you can rebase on it and switch off the reactor stop stuff.

AdamISZ commented:
Done see #660

kristapsk commented:
> I'm not sure why it calls this function twice, once when I first leave the recipient input box, and second time when i click start. that second time, it is not of course a bip21 uri (it's an address) and bi21_uri is set to None, causing the error.

Cannot figure that out by looking at code either, for some reason `editingFinished` on `addressInput` is triggered when you hit Start button. Maybe something Qt-version or GUI environment specific? Ugly hacks could be added to work around this, but I would be more happy to be able to reproduce and fix properly.

kristapsk commented:
> Tbh we should probably take the deferred that is returned by `task.deferLater` and add a callback to it, that wipes the fields; and also disable the fields while the payment is running. I'll try to mock something up later today and see if it more or less works, if so i'll write the code here.

Don't worry, I'm almost finished with that too.

AdamISZ commented:
> Don't worry, I'm almost finished with that too

Got it.

> Maybe something Qt-version or GUI environment specific

I had a suspicion of that also. I did a google for 'editingFinished fired twice Qt' or similar and I think there might have been something, lost that tab now sorry :)

kristapsk commented:
Rebased and addressed review comments. But this is still not perfect. Form fields are disabled and then re-enabled almost immediately, guess need to add some callback when either payjoin or simple direct send tx is broadcast. Also, `accept_callback` and `info_callback` in `send_payjoin()` does not seem to work here either.

jules23 commented on scripts/joinmarket-qt.py:
```diff
@@ -490,25 +514,37 @@ def initUI(self):
+        recipientLabel = QLabel('Recipient address / URI')
         recipientLabel.setToolTip(
-            'The address you want to send the payment to')
+            'The addressi or bitcoin: URI you want to send the payment to')
```
`The address`

kristapsk replied:
Vim thing :D

jules23 replied:
Haha same!

AdamISZ commented on scripts/joinmarket-qt.py:
```diff
@@ -698,9 +735,36 @@ def qt_directsend_callback(rtxd, rtxid, confs):
+            d = task.deferLater(reactor, 0.0, send_payjoin, manager,
+                    accept_callback=self.checkDirectSend,
+                    info_callback=self.infoDirectSend,
+                    tls_whitelist=tls_whitelist_for_testing)
```
Only apply the whitelist when we are on regtest (check `isinstance(jm_single().bc_interface, RegtestBitcoinCoreInterface` iirc), to be safe.

AdamISZ replied:
Wait, I just had another thought about this, this is none of Qt's business; we can put this check in `jmclient.payjoin.py` instead. Even more, we should be putting all blockchain logic code inside the walletservice, since we really want to decouple all the client stuff from anything low level like blockchain for easier changes down the line. So either put this check in payjoin.py via the sendpayment function, and/or have that function query the wallet service for whether we are on regtest or not.

AdamISZ commented:
First tests on https://github.com/JoinMarket-Org/joinmarket-clientserver/commit/1fb06e77424258f560258ce78905720c7670adfd are looking good: `reactor.stop()` bug is fixed, pressing Start did not result in the previously mentioned bug this time (did you figure out what was causing that?) relating to editingFinished firing, payjoin was successful (regtest instance of `test/payjoinserver.py`), and there was no timeout fallback triggered after 60s. Finally, after the tx completed I have the form fields back to normal and emptied.
Just a first test of course, but all good so far.

kristapsk commented:
> pressing Start did not result in the previously mentioned bug this time (did you figure out what was causing that?)

Nop, have no idea.

jules23 commented:
I just executed a couple of payjoins against a testnet remote BTCPay Server instance, with both clearnet and onion urls, and with both nested and native segwit.  It all looks to be working!

The txids for reference:

698280f63dcd2ce5f6e2ff767d02e47a01b237ffad4d92298c8bd3ae1c94ca38
8862c1233c054b76ac59abb19febc98343c2c5f10c7e1716e8c031f3b8482485
0f0a6f9f879277c833544ab596fdda0d302b2ea67f7cb6eadbb26256e4e3a304 - made with an onion url
66aeeec348b8947404e631a19ba9aa775ccec3a85928fc20c24dd2002a7979ea
16da319c547bf0d36f5a3487edf3bc8257e087958fc19eff3c704ed7073440cf
ad730d9a9211d272477ce3431cef36cc3ee952ad6148a6a5ba3e5ff2e9e80d9f - made with an onion url

AdamISZ commented:
> > pressing Start did not result in the previously mentioned bug this time (did you figure out what was causing that?)
> 
> Nop, have no idea.

So, I don't think this is the *full* story, but there's something quite profound here:

The general logic being applied is: if we parse the recipient field as a BIP21 URI, we toggle the fields and populate them, *including changing the content of the recipient field to an address*. Consider then, what happens if the user clicks into the recipient field again, after this `switchToBIP78Payjoin` event has occurred. The field now contains only an address, and when the user *leaves* the field (QLineEdit), the `editingFinished` is fired, the `checkAddress()` is called, and this line gets executed: https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/659/files#diff-3a1ec99e5c3829c412ea7b81a7b15be7R332

As a result, on pressing 'Start' we get an error that the BIP21 URI is invalid.

Potential solution: add a 'Cancel' button that runs `_clearFields` (or similar) so that a user can exit the workflow of the payjoin, while at the same time, disable editing of the fields of the payjoin that have been parsed (I think this is kind of in line with the concept of BIP21, the user shouldn't be editing those fields).

kristapsk commented:
Addressed typo and tls_whitelist review comments so far, UI/UX issues remain TODO.

AdamISZ commented:
So I've been testing today and I think this solution is satisfactory:

(1) move `clearFields` to a member function of `spendTab` to reuse it, and include in it the reset of the Abort button to disabled:

```
    def clearFields(self, ignored):
        self.switchToJoinmarket()
        self.addressInput.setText('')
        self.amountInput.setText('')
        self.addressInput.setEnabled(True)
        self.pjEndpointInput.setEnabled(True)
        self.mixdepthInput.setEnabled(True)
        self.amountInput.setEnabled(True)
        self.startButton.setEnabled(True)
        self.abortButton.setEnabled(False)
```

(2) remove the `pjEndpointEditingFinished` function and the reference to it.

(3) Add two calls in `switchToBIP78Payjoin as below:

```python3
    def switchToBIP78Payjoin(self, endpoint_url):
        self.numCPLabel.setVisible(False)
        self.numCPInput.setVisible(False)
        self.pjEndpointInput.setText(endpoint_url)
        self.pjEndpointLabel.setVisible(True)
        self.pjEndpointInput.setVisible(True)

        # while user is attempting a payjoin, address
        # cannot be edited; to back out, they hit Abort.
        self.addressInput.setEnabled(False)
        self.abortButton.setEnabled(True
```

(4) Finally, have `abortTransactions` call `clearFields` in case payjoin is currently active:

```python3
    def abortTransactions(self):
        if self.pjEndpointInput.isVisible():
            self.clearFields(None)
        elif self.taker:
            self.taker.aborted = True
            self.giveUp()
```

With these alterations i find the behaviour is intuitive and correct; it avoids the above bug (which I was finding was happening *most* of the time), it disables the address edit (while still displaying the address), and activates the 'Abort' option, which when used reverts to the Joinmarket format.

If you're OK with this I can merge this PR as is then add that one commit.

kristapsk commented:
It's ok, but I guess I need to squash commits too. Can do that in a few hours.

AdamISZ commented:
Yes, thanks. I'll merge it tomorrow, assuming there aren't other problems.

kristapsk commented:
Squashed commits, added changes you were proposing + import `RegtestBitcoinCoreInterface` was missing from `payjoin.py`.

AdamISZ commented:
Passes test suite.
Tested on native and p2sh and working correctly (except for a missing 'b' on the tls_whitelist entry that I'll add in a bugfix commit).
tACK and merging.

Doubtless we'll refine the UI a bit more later down the line, especially after more people have tried it, but this works.
